### PR TITLE
Allow PV/Claim matching on inexact access modes

### LINKED ISSUE #10683: Better PV matching
Currently, the PV binder matches claims with the exact access modes on the volume and claim.  For example, a GCEPD with RWO+ROX will bind only to a claim that requests RWO+ROX.

Matching should be loosened to allow users to request fewer modes but possibly get more than they asked for.  A user can request RWO in a claim but get a GCEPD with RWO+ROX.

Getting more than requested follows storage, which matches on at least the amount requested (5Gi) but could be more (the volume actually has 10Gi).

### PR DESCRIPTION
**Pull request by markturansky**
Resolves #10683 

Use case 1:  User requests only ReadWriteOnce.  Cluster is full of GCEPDs (RWO+ROX).  The user's request should match a GCEPD.

Use case 2:  User's claim requests only ReadWriteMany.   There is no exact match for that (NFS is RWO+ROX+RWX).  The request for ReadWriteMany should be satisfied by an NFS volume.

@thockin @smarterclayton

### CODE REVIEW

File: pkg/client/cache/store.go [hunk 1]
```diff
@@ -164,6 +164,11 @@ func (c *cache) Index(indexName string, obj interface{}) ([]interface{}, error)
 	return c.cacheStorage.Index(indexName, obj)
 }
 
+// ListIndexNames returns a list of names used by the Index
+func (c *cache) ListIndexNames(name string) []string {
+	return c.cacheStorage.ListIndexNames(name)
+}
+
```
Comment (markturansky) [inline pkg/client/cache/store.go]: @derekwaynecarr would you mind reviewing this change?  I added the ability to retrieve index names from an Index.

Comment (eikaas) [inline pkg/client/cache/store.go]: This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  tapiohush@gmail.com
    Domain eikaas.no has exceeded the max emails per hour (35/30 (116%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext3.iad.github.net ([192.30.252.194]:46930 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by cpanel30.proisp.no with esmtps (TLSv1.2:DHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.85)
	(envelope-from <noreply@github.com>)
	id 1ZEhbn-0021SB-Ic
	for robin@eikaas.no; Mon, 13 Jul 2015 19:33:43 +0200
Date: Mon, 13 Jul 2015 10:33:38 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1436808818;
	bh=0CQ7jlmhKA5ADHnb+WKEXaPcdVqq88Ygrv0CpS05R80=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=GidexJiuUpY2dojh+l+lDLixoI+9eexgUgjsKXH2bmrNdnS2YYtKaCxLzktyg0/kl
	 6eoci9NzR0UEUTRoFMCc4atctTq26M7pxvLb5ObFO/DZrIW835pnezA5Qz4LWSx+zp
	 hHHloFjZXhzfWK4o4QwHIAh4tEqy3ErXey8S6HeA=

File: pkg/client/cache/index.go [hunk 2]
```diff
@@ -28,6 +28,8 @@ type Indexer interface {
 	Store
 	// Retrieve list of objects that match on the named indexing function
 	Index(indexName string, obj interface{}) ([]interface{}, error)
+	// Retrieve list of all index names
+	ListIndexNames(name string) []string
```
Comment (derekwaynecarr) [inline pkg/client/cache/index.go]: This should not take a ```name``` argument.  The idea is to basically give all the keys in the type ```Indexers``` so I do not see why a name argument is provided.

File: pkg/client/cache/store.go [hunk 3]
```diff
@@ -164,6 +164,11 @@ func (c *cache) Index(indexName string, obj interface{}) ([]interface{}, error)
 	return c.cacheStorage.Index(indexName, obj)
 }
 
+// ListIndexNames returns a list of names used by the Index
+func (c *cache) ListIndexNames(name string) []string {
+	return c.cacheStorage.ListIndexNames(name)
```
Comment (derekwaynecarr) [inline pkg/client/cache/store.go]: drop name

File: pkg/client/cache/index.go (hunk 2 above)
Comment (derekwaynecarr) [inline pkg/client/cache/index.go]: Sorry, I think I got confused here.  Is there a test case that you can author to show your desired use of this new function more clearly?  I would have thought you are just looking to get the keys on the ```Indices``` object so name is not needed.  Let me know if I misunderstood something here, but this change seems wrong.

Comment (markturansky) [inline pkg/client/cache/index.go]: When PVs get indexed, their access modes are converted into a string and that string is the index.   Different modes render a different string.

A user makes a PVClaim with some number of access modes that might not match the indexed string (e.g, they request just ReadWriteMany.   This matches NFS (RWO+ROX+RWX).    

I asked the index for all "accessmodes" (the name arg) and from there I get the keys in the index, which are string representations of the access modes of all volumes.    I can then find a RWX mode among the many keys and give the user a match.

persistent_volume_index_test.go, TestFindingVolumeWithDifferentAccessModes is the main test case for this PR.

Comment (derekwaynecarr) [inline pkg/client/cache/index.go]: So is this ```ListIndexKeys```??

Comment (derekwaynecarr) [inline pkg/client/cache/index.go]: I still would like a test case that is not tied to TestFindingBlahBlahBlah

Comment (markturansky) [inline pkg/client/cache/index.go]: Yes and will do.  Also will break this into a separate PR.

Comment (markturansky) [inline pkg/client/cache/index.go]: https://github.com/GoogleCloudPlatform/kubernetes/pull/11171

**Comment by markturansky** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/pull/11171 contains the changes to the Index that allows retrieval of the generated IndexFunc values.   Those changes will be removed from this PR after that PR is merged.

@smarterclayton PTAL?  This was deemed a priority for OpenShift.

**Comment by markturansky** [comment]
@smarterclayton @thockin This is ready for review.   

I rebased and pulled in #11171.

File: pkg/volume/util.go [hunk 4]
```diff
@@ -47,6 +50,21 @@ func GetAccessModesAsString(modes []api.PersistentVolumeAccessMode) string {
 	return modesAsString
 }
 
+func GetAccessModesFromString(modes string) []api.PersistentVolumeAccessMode {
```
Comment (smarterclayton) [inline pkg/volume/util.go]: Seems like these are part of the API, not volumes.  Why not move these to api/helpers.go?

File: pkg/volumeclaimbinder/types.go [hunk 5]
```diff
@@ -75,23 +76,29 @@ type matchPredicate func(compareThis, toThis *api.PersistentVolume) bool
 
 // Find returns the nearest PV from the ordered list or nil if a match is not found
 func (pvIndex *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matchPredicate matchPredicate) (*api.PersistentVolume, error) {
-	volumes, err := pvIndex.ListByAccessModes(pv.Spec.AccessModes)
-	if err != nil {
-		return nil, err
-	}
+	// the user may ask for access modes which either don't exist exactly as specified or can be matched by another mode.
+	// e.g, users asks for ReadWriteOnce but a GCEPD is available, which is ReadWriteOnce+ReadOnlyMay.
```
Comment (smarterclayton) [inline pkg/volumeclaimbinder/types.go]: nit: typo in comment `ReadOnlyMay`

File: pkg/volumeclaimbinder/types.go [hunk 6]
```diff
@@ -139,3 +146,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
...
+//  RWO -- the most basic form of access (all storage volumes)
+//  RWO+ROX -- 1 writer or many read-only readers (GCEPersistentDisk)
+//  RWO+ROX+RWX -- many writers (NFS)
+//
+// A user's request can be satisfied with more modes, but never fewer.
+// This func returns modes with ascending levels of modes to give the user what is closest to what they actually asked for.
+//
+func (pvIndex *persistentVolumeOrderedIndex) allPossibleMatchingAccessModes(requestedModes []api.PersistentVolumeAccessMode) [][]api.PersistentVolumeAccessMode {
+	matchedModes := [][]api.PersistentVolumeAccessMode{}
+	keys := pvIndex.Indexer.ListIndexFuncValues("accessmodes")
```
Comment (smarterclayton) [inline pkg/volumeclaimbinder/types.go]: Why does this value need to come from the indexer?  Is the list of access modes not finite?

Comment (markturansky) [inline pkg/volumeclaimbinder/types.go]: I wanted to account for all PVs that were actually added instead of attempting to derive all possible combinations myself.  Asking for what it has indexed was the easiest way to find all current access modes in use.

Comment (smarterclayton) [inline pkg/volumeclaimbinder/types.go]: Ok

On Fri, Jul 24, 2015 at 11:29 AM, Mark Turansky <notifications@github.com>
wrote:

> In pkg/volumeclaimbinder/types.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10833#discussion_r35433497>
> :
>
> > @@ -139,3 +146,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
> >  	bSize := bQty.Value()
> >  	return aSize <= bSize
> >  }
> > +
> > +// allPossibleMatchingAccessModes returns an array of AccessMode arrays that can satisfy a user's requested modes.
> > +// There are, essentially, 3 ways to request access modes:
> > +//
> > +//  RWO -- the most basic form of access (all storage volumes)
> > +//  RWO+ROX -- 1 writer or many read-only readers (GCEPersistentDisk)
> > +//  RWO+ROX+RWX -- many writers (NFS)
> > +//
> > +// A user's request can be satisfied with more modes, but never fewer.
> > +// This func returns modes with ascending levels of modes to give the user what is closest to what they actually asked for.
> > +//
> > +func (pvIndex *persistentVolumeOrderedIndex) allPossibleMatchingAccessModes(requestedModes []api.PersistentVolumeAccessMode) [][]api.PersistentVolumeAccessMode {
> > +	matchedModes := [][]api.PersistentVolumeAccessMode{}
> > +	keys := pvIndex.Indexer.ListIndexFuncValues("accessmodes")
>
> I wanted to account for all PVs that were actually added instead of
> attempting to derive all possible combinations myself. Asking for what it
> has indexed was the easiest way to find all current access modes in use.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10833/files#r35433497>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

File: pkg/volume/util.go (hunk 4 above)
Comment (markturansky) [inline pkg/volume/util.go]: This is a good suggestion.  It ended with callers of these func having fewer imports.  Less is less.

**Comment by markturansky** [comment]
@smarterclayton feedback addressed.

**Comment by smarterclayton** [comment]
I don't have other issues with this code, but I'm not an expert reviewer on the volume claim binder so defer to tim.

**Comment by markturansky** [comment]
@thockin PTAL?

File: pkg/api/helpers.go [hunk 7]
```diff
@@ -178,3 +179,53 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
+	for _, s := range strmodes {
+		if s == "RWO" {
+			accessModes = append(accessModes, ReadWriteOnce)
+		} else if s == "ROX" {
+			accessModes = append(accessModes, ReadOnlyMany)
+		} else if s == "RWX" {
+			accessModes = append(accessModes, ReadWriteMany)
+		}
+	}
+	return accessModes
```
Comment (thockin) [inline pkg/api/helpers.go]: This returns an unsorted list - is that OK?  It also could have dup values if bad input.  Is that OK?

You could (totally optional) unpack into a StringSet, then explode that into a list.

File: pkg/api/helpers.go [hunk 8]
```diff
@@ -178,3 +179,53 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
 	return c
 }
+
+// GetAccessModesAsString returns a string representation of an array of access modes.
+// modes, when present, are always in the same order: RWO,ROX,RWX.
+func GetAccessModesAsString(modes []PersistentVolumeAccessMode) string {
+	modesAsString := ""
+
+	if contains(modes, ReadWriteOnce) {
+		appendAccessMode(&modesAsString, "RWO")
```
Comment (thockin) [inline pkg/api/helpers.go]: Why not keep a map[string]string that maps api.PersistentVolumeAccessMode -> short-string and vice-versa, then just iterate here and below.  That way you only spell out the full list one time.  You just KNOW that one day there will be a new mode and you will have to hunt down 75 places you hardcoded these.

File: pkg/api/helpers.go [hunk 9]
```diff
@@ -178,3 +179,53 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
+}
+
+func appendAccessMode(modes *string, mode string) {
+	if *modes != "" {
+		*modes += ","
+	}
+	*modes += mode
+}
+
+func contains(modes []PersistentVolumeAccessMode, mode PersistentVolumeAccessMode) bool {
```
Comment (thockin) [inline pkg/api/helpers.go]: better name?  This is exposed to all of pkg api, since Go doesn't have file-static exposure.

File: pkg/volumeclaimbinder/types.go [hunk 10]
```diff
@@ -139,3 +145,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
 	bSize := bQty.Value()
 	return aSize <= bSize
 }
+
+// allPossibleMatchingAccessModes returns an array of AccessMode arrays that can satisfy a user's requested modes.
+// There are, essentially, 3 ways to request access modes:
+//
+//  RWO -- the most basic form of access (all storage volumes)
+//  RWO+ROX -- 1 writer or many read-only readers (GCEPersistentDisk)
+//  RWO+ROX+RWX -- many writers (NFS)
```
Comment (thockin) [inline pkg/volumeclaimbinder/types.go]: I was operating under the assumption that a use requested RWX and that implied RWO and ROX.  Is that not the case?  Does a user have to actually spell out modes they don't want?

File: pkg/volumeclaimbinder/types.go [hunk 11]
```diff
@@ -75,23 +75,29 @@ type matchPredicate func(compareThis, toThis *api.PersistentVolume) bool
 
 // Find returns the nearest PV from the ordered list or nil if a match is not found
 func (pvIndex *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matchPredicate matchPredicate) (*api.PersistentVolume, error) {
-	volumes, err := pvIndex.ListByAccessModes(pv.Spec.AccessModes)
-	if err != nil {
-		return nil, err
-	}
+	// the user may ask for access modes which either don't exist exactly as specified or can be matched by another mode.
```
Comment (thockin) [inline pkg/volumeclaimbinder/types.go]: This comment talks about a user requesting modes, but pv.Spec is the OFFERED modes, isn't it?  This could be clearer.

File: pkg/volumeclaimbinder/types.go [hunk 12]
```diff
@@ -139,3 +145,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
...
 }
+
+// allPossibleMatchingAccessModes returns an array of AccessMode arrays that can satisfy a user's requested modes.
+// There are, essentially, 3 ways to request access modes:
+//
+//  RWO -- the most basic form of access (all storage volumes)
+//  RWO+ROX -- 1 writer or many read-only readers (GCEPersistentDisk)
+//  RWO+ROX+RWX -- many writers (NFS)
+//
+// A user's request can be satisfied with more modes, but never fewer.
```
Comment (thockin) [inline pkg/volumeclaimbinder/types.go]: Same comment as above - do you expect this func to receive the requested mode: "I need at least ROX" or do you expect to receive the offered mode: "I can be used as ROX".  The former expands to [ ROX, RWX ], but the latter expands to [  RWO, ROX ].  From the call site this looks to be the offered mode, but the comment disagrees.

File: pkg/volumeclaimbinder/types.go [hunk 13]
```diff
@@ -139,3 +145,65 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
...
+// This func returns modes with ascending levels of modes to give the user what is closest to what they actually asked for.
```
Comment (thockin) [inline pkg/volumeclaimbinder/types.go]: Some examples of input -> output would also help.  I don't grok slice-of-slice-of mode as the return

File: pkg/api/helpers.go (hunk 8 above)
Comment (markturansky) [inline pkg/api/helpers.go]: The order would not be determinant if I used a map.   The string is conveniently in ascending order of access.

File: pkg/api/helpers.go (hunk 9 above)
Comment (markturansky) [inline pkg/api/helpers.go]: renamed.

**Comment by markturansky** [comment]
@thockin re-written comments, hopefully they are much clearer and address your points made above.

File: pkg/api/helpers.go (hunk 8 above)
Comment (markturansky) [inline pkg/api/helpers.go]: Mean to put that comment above :)

The only place I'm using these constants (outside of tests) is here in helpers for serialization to/from a string.   The binder uses the string for indexing.  The CLI uses the string, too.

**Comment by markturansky** [comment]
Rebased and squashed for easier review.

**Comment by markturansky** [comment]
@thockin PTAL?  this is ready for review.

File: pkg/api/helpers.go [hunk 14]
```diff
@@ -179,3 +180,65 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
+
+	return modesAsString
+}
+
+// GetAccessModesAsString returns an array of AccessModes from a string created by GetAccessModesAsString
+func GetAccessModesFromString(modes string) []PersistentVolumeAccessMode {
+	strmodes := strings.Split(modes, ",")
+	accessModes := []PersistentVolumeAccessMode{}
+	for _, s := range strmodes {
+		if s == "RWO" {
```
Comment (thockin) [inline pkg/api/helpers.go]: might want to Trim(s, " ") ?  or are we sure no whitespace will creep in?

File: pkg/api/helpers.go (hunk 7 above)
Comment (thockin) [inline pkg/api/helpers.go]: This comment is not addressed.  Is it OK that "RWX,RWO,ROX" produces a different-but-equivalent slice as "ROX,RWX,RWO" ?  Is slice the right structure, or should it be a Set?

File: pkg/api/helpers.go [hunk 15]
```diff
@@ -179,3 +180,65 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
+
+	if containsAccessMode(modes, ReadWriteOnce) {
+		appendAccessMode(&modesAsString, "RWO")
+	}
+	if containsAccessMode(modes, ReadOnlyMany) {
+		appendAccessMode(&modesAsString, "ROX")
+	}
+	if containsAccessMode(modes, ReadWriteMany) {
+		appendAccessMode(&modesAsString, "RWX")
+	}
```
Comment (thockin) [inline pkg/api/helpers.go]: simpler to do:

```
modes := []string{}
if containsAccessMode(modes, ReadWriteOnce) {
		modes = append(modes, "RWO")
}
if containsAccessMode(modes, ReadOnlyMany) {
		modes = append(modes, "ROX")
}
if containsAccessMode(modes, ReadWriteMany) {
		modes = append(modes, "RWX")
}
return strings.Join(modes, ",")

File: pkg/api/helpers.go [hunk 16]
```diff
@@ -179,3 +180,65 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
+func appendAccessMode(modes *string, mode string) {
+	if *modes != "" {
+		*modes += ","
+	}
+	*modes += mode
+}
+
+// removeDuplicateAccessModes returns an array of access modes without any duplicates
+func removeDuplicateAccessModes(modes []PersistentVolumeAccessMode) []PersistentVolumeAccessMode {
+	accessModes := []PersistentVolumeAccessMode{}
```
Comment (thockin) [inline pkg/api/helpers.go]: This whole function can be replace with `return util.NewStringSet(modes...).List()`

File: pkg/api/helpers.go [hunk 17]
```diff
@@ -179,3 +180,65 @@ func LoadBalancerStatusDeepCopy(lb *LoadBalancerStatus) *LoadBalancerStatus {
...
+	for _, m := range modes {
+		if !containsAccessMode(accessModes, m) {
+			accessModes = append(accessModes, m)
+		}
+	}
+	return accessModes
+}
+
+func containsAccessMode(modes []PersistentVolumeAccessMode, mode PersistentVolumeAccessMode) bool {
```
Comment (thockin) [inline pkg/api/helpers.go]: I have to ask again - is slice the right abstraction for passing this around?  Or should we extract it from API into a StringSet and use that all over?

File: pkg/controller/persistentvolume/types.go [hunk 18]
```diff
@@ -75,23 +74,38 @@ type matchPredicate func(compareThis, toThis *api.PersistentVolume) bool
...
+	//
+	// PVs are indexed by their access modes to allow easier searching.  Each index is the string representation of a set of access modes.
+	// There is a finite number of possible sets and PVs will only be indexed in one of them (whichever index matches the PV's modes).
+	//
+	// A request for resources will always specify its desired access modes.  Any matching PV must have at least that number
+	// of access modes, but it can have more.  For example, a user asks for ReadWriteOnce but a GCEPD is available, which is ReadWriteOnce+ReadOnlyMany.
+	//
+	// Searches are performed against a set of access modes, so we can attempt not only the exact matching modes but also
+	// potential matches (the GCEPD example above).
+	allPossibleModes := pvIndex.allPossibleMatchingAccessModes(pv.Spec.AccessModes)
```
Comment (thockin) [inline pkg/controller/persistentvolume/types.go]: I'm not sure I understand why this is two index calls.

1) get a list of lists of modes that satisfy
2) get a list of volumes with each list

why can't it just get a list of volumes that satisfy in a single call to the index?

File: pkg/controller/persistentvolume/types.go [hunk 19]
```diff
@@ -139,3 +153,88 @@ func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
...
+		if containedInAll(indexedModes, requestedModes) {
+			matchedModes = append(matchedModes, indexedModes)
+		}
+	}
+
+	// sort by the number of modes in each array with the fewest number of modes coming first.
+	// this allows searching for volumes by the minimum number of modes required of the possible matches.
+	sort.Sort(byAccessModes{matchedModes})
+	return matchedModes
+
```
Comment (thockin) [inline pkg/controller/persistentvolume/types.go]: dead line (you asked for it)

**Comment by thockin** [comment]
@saad-ali

File: pkg/api/helpers.go (hunk 17 above)
Comment (pmorie) [inline pkg/api/helpers.go]: I'm in favor of converting into strings internally and working with them that way.  The typedef in the API is mainly about validating that you get the correct values, is it not?

Comment (markturansky) [inline pkg/api/helpers.go]: Is order guaranteed in a StringSet?  As a Set, I presume not.

PVs are indexed once -- under their stringified access modes.  The helper funcs always create consistent strings for indexing those PVs.

File: pkg/api/helpers.go (hunk 16 above)
Comment (markturansky) [inline pkg/api/helpers.go]: Can't because of typing.  I would need a loop anyway.

`cannot use modes (type []PersistentVolumeAccessMode) as type []string in argument to util.NewStringSet`

File: pkg/api/helpers.go (hunk 15 above)
Comment (markturansky) [inline pkg/api/helpers.go]: changed to this suggestion.

File: pkg/api/helpers.go (hunk 14 above)
Comment (markturansky) [inline pkg/api/helpers.go]: Added Trim, just in case.  Because stringifying happened in the helper, I wasn't worried about decoding that same string without Trim.  Can't hurt, though.

File: pkg/controller/persistentvolume/types.go (hunk 18 above)
Comment (markturansky) [inline pkg/controller/persistentvolume/types.go]: This was, to me, the easiest way to search.

Each PV is indexed exactly once using a string key that is the PV's stringified access modes.

Different sets of access modes might match what the user asked for, but the index beneath the distinct stringified access modes might not contain the PV.  

This code calls into the index for each, as you noted.  Changing how the index keeps volumes and searches for them in a bigger change that simply dipping into the index more than once to find a match.

**Comment by markturansky** [comment]
@thockin there were a few things addressed here in comments, a few others were changed in the code.  I believe this PR is done, because changing how the index works would be a bigger endeavor.

**Comment by saad-ali** [comment]
I'll take a look at this today

**Comment by markturansky** [comment]
Thanks @saad-ali.   The only comment of Tim's not addressed is searching the index, but I left comments above in explanation.  Otherwise, all feedback is accounted for and we've got all green lights.

File: pkg/controller/persistentvolume/types.go (hunk 18 above)
Comment (saad-ali) [inline pkg/controller/persistentvolume/types.go]: This code is pretty hard to follow, mostly because of the multiple levels of indirection: a mapping from a fixed key (`accessmodes`) to an indexing function that takes a PV object and spits out an array of strings of length one which is a concatenation of modes in a specific order.

I'd love to see this simplified, if at all possible.

Without fully understanding all the constraints you're working with, how about something like maintaining a max heaps of PVs (sorted by storage capacity) for each possible mode (RWO, ROX, RWX), where each PV could be in multiple heaps depending on the modes they support?

Comment (markturansky) [inline pkg/controller/persistentvolume/types.go]: @saad-ali Is there a way to make this current implementation more clear?  

What you are proposing is a rewrite of the PV index.  That is, to me, riskier and more time consuming than this issue warrants.

I think, too, there is an equivalence here.

1.  Figure out all the possible combinations of access modes a PV can have (or are requested) and index the PV in each.  This allows us to query the index once.

2.  Figure out all possible combinations that satisfy a request for a PV and query the index for each.  Keep the PV in a single index.

Either way, It seems like one piece has to go from 1 to Many.  I choose the second option because it was simpler than trying to rewrite the index, which is riskier, IMHO.

Thoughts?

Comment (saad-ali) [inline pkg/controller/persistentvolume/types.go]: @markturansky That's fair. I don't see a way of easily simplifying this short of rewriting the indexing code and that's out-of-scope for this PR.

**Comment by saad-ali** [comment]
LGTM

**Comment by saad-ali** [comment]
@k8s-bot ok to test
